### PR TITLE
Switch ui.window.devicePixelRatio from browser logical to physical.

### PR DESCRIPTION
**Pull request by ferhatb**
- This enabled image loading for multi dpi assets and fixes correctness issues in framework due to incorrectly using logical browser pixels at  1.0 dpr.
- Adds integration test for image loading of higher resolution.
- Adds missing platform replymessage calls.

Related issues:
https://github.com/flutter/flutter/issues/46186
https://github.com/flutter/flutter/issues/41328
https://github.com/material-components/material-components-flutter-gallery/issues/310

Fixes: https://github.com/flutter/flutter/issues/52929

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -146,6 +146,7 @@ task:
         - $FRAMEWORK_PATH/flutter/bin/flutter config --local-engine=host_debug_unopt --no-analytics --enable-web
         - $FRAMEWORK_PATH/flutter/bin/flutter pub get --local-engine=host_debug_unopt
         - $FRAMEWORK_PATH/flutter/bin/flutter drive -v --target=test_driver/text_editing_e2e.dart -d web-server --release --browser-name=chrome --local-engine=host_debug_unopt
+        - $FRAMEWORK_PATH/flutter/bin/flutter drive -v --target=test_driver/image_loading_e2e.dart -d web-server --release --browser-name=chrome --local-engine=host_debug_unopt
```
Comment (nturgut) [inline .cirrus.yml]: soon I'll run them all from felt :)

File: lib/web_ui/lib/src/engine/surface/scene_builder.dart [hunk 2]
```diff
@@ -97,6 +97,14 @@ class SurfaceSceneBuilder implements ui.SceneBuilder {
     if (matrix4.length != 16) {
       throw ArgumentError('"matrix4" must have 16 entries.');
     }
+    if (_surfaceStack.length == 1) {
+      // Top level transform contains view configuration to scale
+      // scene to devicepixelratio. Use identity instead since web browser
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/surface/scene_builder.dart]: nit: to be more specific CSS uses logical pixels (e.g. WebGL doesn't).

File: lib/web_ui/lib/src/engine/window.dart [hunk 3]
```diff
@@ -36,12 +28,10 @@ class EngineWindow extends ui.Window {
...
-  /// This is useful in tests to emulate screens of different dimensions.
-  void debugOverrideDevicePixelRatio(double value) {
-    assert(() {
-      _debugDevicePixelRatio = value;
-      return true;
-    }());
+  /// This is useful in tests to emulate screens of different dimensions
+  /// including integration tests.
+  void overrideDevicePixelRatio(double value) {
+    _debugDevicePixelRatio = value;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart]: Why this change? It's still used only in tests, so the "debug" prefix and `assert` seem appropriate.

File: e2etests/web/regular_integration_tests/test_driver/image_loading_e2e_test.dart [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import 'dart:io';
+
+import 'package:flutter_driver/flutter_driver.dart';
+
+Future<void> main() async {
+  final FlutterDriver driver = await FlutterDriver.connect();
+
+  final String dataRequest =
+      await driver.requestData(null, timeout: const Duration(seconds: 1));
+  print('result $dataRequest');
```
Comment (yjbanov) [inline e2etests/web/regular_integration_tests/test_driver/image_loading_e2e_test.dart]: remove the print?

File: .cirrus.yml (hunk 1 above)
Comment (ferhatb) [reply .cirrus.yml]: 👍

File: lib/web_ui/lib/src/engine/window.dart (hunk 3 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/window.dart]: It is now also used in integration test (the one that loads an image at framework level).

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/window.dart]: Right, but that's still a test :)

File: lib/web_ui/lib/src/engine/surface/scene_builder.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/scene_builder.dart]: Done.

File: e2etests/web/regular_integration_tests/test_driver/image_loading_e2e_test.dart (hunk 4 above)
Comment (ferhatb) [reply e2etests/web/regular_integration_tests/test_driver/image_loading_e2e_test.dart]: Done.

File: lib/web_ui/lib/src/engine/window.dart [hunk 5]
```diff
@@ -182,6 +172,9 @@ class EngineWindow extends ui.Window {
         textEditing.channel.handleTextInput(data);
         return;
 
+      case 'flutter/web_test_e2e':
+        _handleWebTestEnd2EndMessage(data);
+        return;
```
Comment (nturgut) [inline lib/web_ui/lib/src/engine/window.dart]: Does testing with mobile emulation ok? Looks like there is an option for setting the device pixel ratio for chrome driver:http://chromedriver.chromium.org/mobile-emulation 
Currently we don't have a way of passing options to the 'flutter drive' command. 

Can you please add a todo? I can remove this part once I added options to chrome driver :)

Comment (ferhatb) [reply lib/web_ui/lib/src/engine/window.dart]: We would need to be able to pass all other browsers as well for integration tests. There are issues floating around from 2018 still not fixed. We may have to keep this for some browsers for a while. For Chrome we could use option, for other browsers use this channel.

Comment (nturgut) [reply lib/web_ui/lib/src/engine/window.dart]: Thanks for the clarification.

File: e2etests/web/regular_integration_tests/lib/image_loading_main.dart [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+  @override
+  void initState() {
+    super.initState();
+    const MethodChannel channel =
+        OptionalMethodChannel('flutter/web_test_e2e', JSONMethodCodec());
+    channel.invokeMethod<void>(
+      'setDevicePixelRatio',
+      '1.5',
+    );
```
Comment (mdebbar) [inline e2etests/web/regular_integration_tests/lib/image_loading_main.dart]: Part of me says this is working today only because platform messages are synchronous on web. If platform messages ever become async in web, this won't work anymore.

Maybe make `main()` async, and send-and-await this platform message before `runApp()`?

File: lib/web_ui/lib/src/engine/window.dart (hunk 3 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/window.dart]: Do integration tests strip out asserts?

File: lib/web_ui/lib/src/engine/window.dart (hunk 5 above)
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/window.dart]: One day, someone should go through all the platform messages and call `callback` which indicates to the framework that the handling of the message is done.

For now, you may need to do it for this platform message by calling:
```dart
_replyToPlatformMessage(callback, codec.encodeSuccessEnvelope(true));
```
This will let you await on the `channel.invokeMethod()` in the framework side.

File: lib/web_ui/lib/src/engine/window.dart (hunk 3 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/window.dart]: If they don't, we should add the option to enable them. `dart2js` supports asserts (we use `dart2js` for engine unit-test with assertions enabled).

/cc @nturgut

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/window.dart]: My comment is non-blocking though. I don't think there's major risk. It just doesn't feel right for a method that's not annotated as "debug" to manipulate a "debug" field without an assertion.

File: e2etests/web/regular_integration_tests/lib/image_loading_main.dart (hunk 6 above)
Comment (yjbanov) [reply e2etests/web/regular_integration_tests/lib/image_loading_main.dart]: +1. Good catch, Mouad. This could cause race conditions down the line. I'm wondering if this falls under the same category as the experiments API, where we use JS-interop instead.

File: lib/web_ui/lib/src/engine/window.dart (hunk 5 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/window.dart]: Filed https://github.com/flutter/flutter/issues/52929 so we don't forget.

**Comment by yjbanov** [comment]
LGTM if LGT @mdebbar

File: e2etests/web/regular_integration_tests/lib/image_loading_main.dart (hunk 6 above)
Comment (ferhatb) [reply e2etests/web/regular_integration_tests/lib/image_loading_main.dart]: Done. Thank you for pointing out reply issue for async.

File: lib/web_ui/lib/src/engine/window.dart (hunk 5 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/window.dart]: Done.

Comment (mdebbar) [reply lib/web_ui/lib/src/engine/window.dart]: Thanks for doing this @ferhatb! Last time I tried, it broke things. I'm glad nothing is broken anymore.

File: e2etests/web/regular_integration_tests/lib/text_editing_main.dart [hunk 7]
```diff
@@ -28,7 +28,7 @@ class MyHomePage extends StatefulWidget {
 
 class _MyHomePageState extends State<MyHomePage> {
   final TextEditingController _controller =
-      TextEditingController(text: 'Text1');
+  TextEditingController(text: 'Text1');
```
Comment (yjbanov) [inline e2etests/web/regular_integration_tests/lib/text_editing_main.dart]: bad indent

File: lib/web_ui/lib/src/engine/semantics/accessibility.dart [hunk 8]
```diff
@@ -52,9 +52,10 @@ class AccessibilityAnnouncements {
   html.HtmlElement get _domElement => _element ??= _createElement();
 
   /// Decodes the message coming from the 'flutter/accessibility' channel.
-  void handleMessage(ByteData data) {
+  StandardMessageCodec handleMessage(ByteData data) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/semantics/accessibility.dart]: I'd prefer to keep this method `void` and have it accept a message codec as an argument.

File: e2etests/web/regular_integration_tests/lib/text_editing_main.dart (hunk 7 above)
Comment (ferhatb) [reply e2etests/web/regular_integration_tests/lib/text_editing_main.dart]: Done.

File: lib/web_ui/lib/src/engine/semantics/accessibility.dart (hunk 8 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/semantics/accessibility.dart]: Done.

File: lib/web_ui/lib/src/engine/surface/scene_builder.dart [hunk 9]
```diff
@@ -97,6 +97,14 @@ class SurfaceSceneBuilder implements ui.SceneBuilder {
     if (matrix4.length != 16) {
       throw ArgumentError('"matrix4" must have 16 entries.');
     }
+    if (_surfaceStack.length == 1) {
+      // Top level transform contains view configuration to scale
+      // scene to devicepixelratio. Use identity instead since CSS uses
+      // logical device pixels.
+      assert(matrix4[0] == EngineWindow.browserDevicePixelRatio &&
+          matrix4[5] == EngineWindow.browserDevicePixelRatio);
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/scene_builder.dart]: Removed the check for test environment since there are many SceneBuilder raw tests outside the context of an app that pushes root transform for now (we really need a ui.Window level debugDevicePixelRatio, i noticed some hacked in TestWindow.devicePixelRatioTestValue etc in the code base used for viewport_test).

File: lib/web_ui/lib/src/engine/surface/scene_builder.dart [hunk 10]
```diff
@@ -97,6 +100,16 @@ class SurfaceSceneBuilder implements ui.SceneBuilder {
...
     }
+    if (_surfaceStack.length == 1) {
+      // Top level transform contains view configuration to scale
+      // scene to devicepixelratio. Use identity instead since CSS uses
+      // logical device pixels.
+      if (!ui.debugEmulateFlutterTesterEnvironment) {
+        assert(matrix4[0] == window.devicePixelRatio &&
+           matrix4[5] == window.devicePixelRatio);
+      }
+      matrix4 = Matrix4.identity().storage;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/surface/scene_builder.dart]: Actually, now that I think about it, the correct fix would be to inject a transform that scales by 1/DPR. This way we don't have to rely on a particular framework behavior. To avoid extra div nodes we can do something like:

```
if (the scene contains a root transform) {
  reuse the root transform by scale it by 1/DPR.
} else {
  apply 1/DPR to the `<flt-scene>` element.
}
```

In most cases we'll hit the `if` branch, not the `else` branch. Also in most cases the root transform will be scale by DPR. So the resulting root transform will automatically become identity and we won't apply any transform at all.

File: lib/web_ui/lib/src/engine/window.dart [hunk 11]
```diff
@@ -15,33 +15,23 @@ class EngineWindow extends ui.Window {
...
-
-  /// Returns device pixel ratio returns by browser.
+  /// Returns device pixel ratio returned by browser.
   static double get browserDevicePixelRatio {
     double ratio = html.window.devicePixelRatio;
     // Guard against WebOS returning 0.
     return (ratio == null || ratio == 0.0) ? 1.0 : ratio;
   }
 
+
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/window.dart]: revert?

File: lib/web_ui/lib/src/ui/test_embedding.dart [hunk 12]
```diff
@@ -29,17 +29,21 @@ Future<dynamic> ensureTestPlatformInitializedThenRunTest(
...
   // Force-initialize DomRenderer so it doesn't overwrite test pixel ratio.
   engine.domRenderer;
 
-  // The following parameters are hard-coded in Flutter's test embedder. Since
-  // we don't have an embedder yet this is the lowest-most layer we can put
-  // this stuff in.
-  engine.window.debugOverrideDevicePixelRatio(devicePixelRatio);
-  engine.window.webOnlyDebugPhysicalSizeOverride =
-      Size(800 * devicePixelRatio, 600 * devicePixelRatio);
+  assert(() {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/ui/test_embedding.dart]: Why is this assert necessary? This code emulates the `flutter test` environment (it sets `debugEmulateFlutterTesterEnvironment` to true below). So it shouldn't affect e2e tests. Those should not hit this code.

File: lib/web_ui/test/golden_tests/engine/compositing_golden_test.dart [hunk 13]
```diff
@@ -405,6 +405,11 @@ void _testCullRectComputation() {
   // enough to fit the rotated clip.
   test('clips correctly when using 3d transforms', () async {
     final SurfaceSceneBuilder builder = SurfaceSceneBuilder();
+
+    builder.pushTransform(Matrix4.diagonal3Values(
+        EngineWindow.browserDevicePixelRatio,
+        EngineWindow.browserDevicePixelRatio, 1.0).storage);
```
Comment (yjbanov) [inline lib/web_ui/test/golden_tests/engine/compositing_golden_test.dart]: This change wouldn't be necessary if we applied 1/DPR to `<flt-scene>` when the root transform is missing.

I think some of our benchmarks will fail for the same reason.

File: lib/web_ui/lib/src/engine/semantics/accessibility.dart [hunk 14]
```diff
@@ -52,9 +52,9 @@ class AccessibilityAnnouncements {
   html.HtmlElement get _domElement => _element ??= _createElement();
 
   /// Decodes the message coming from the 'flutter/accessibility' channel.
-  void handleMessage(ByteData data) {
+   void handleMessage(StandardMessageCodec codec, ByteData data) {
```
[the remark this answers is not in the document]
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/semantics/accessibility.dart]: Done.

File: lib/web_ui/lib/src/engine/surface/scene_builder.dart (hunk 10 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/surface/scene_builder.dart]: The framework itself injects pushTransform using ViewConfiguration. Ideally this code would not live in framework but in engine. See view.dart:39. Or instead of calling pushTransform , framework would call SceneBuilder.setViewConfiguration in engine.

File: lib/web_ui/lib/src/engine/window.dart (hunk 11 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/window.dart]: Done.

File: lib/web_ui/lib/src/ui/test_embedding.dart (hunk 12 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/ui/test_embedding.dart]: reverted

File: lib/web_ui/test/golden_tests/engine/compositing_golden_test.dart (hunk 13 above)
Comment (ferhatb) [reply lib/web_ui/test/golden_tests/engine/compositing_golden_test.dart]: Agreed. Would like to change ViewConfiguration so we don't have to do this in the first place on the framework side.

File: lib/web_ui/lib/src/engine/surface/scene_builder.dart (hunk 10 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/surface/scene_builder.dart]: Right, but you can use `dart:ui` without the Flutter framework where you can paint a scene without a root transform. This typically happens in tests and benchmarks. I don't expect there to be many apps that don't use the framework.

**Comment by ferhatb** [comment]
merging on luci-engine infra fail.
